docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
  return abcitypes.ResponseOfferSnapshot{}
 }
 
-func (KVStoreApplication) LoadSnapshotChunk(abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
+func (app *KVStoreApplication) LoadSnapshotChunk(abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
  return abcitypes.ResponseLoadSnapshotChunk{}
 }
 
-func (KVStoreApplication) ApplySnapshotChunk(abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
+func (app *KVStoreApplication) ApplySnapshotChunk(abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
  return abcitypes.ResponseApplySnapshotChunk{}
 }
 ```
 
-Now I will go through each method explaining when it's called and adding
-required business logic.
+### 1.3.1 Add a persistent data store
 
-### 1.3.1 CheckTx
+Our application will need to write its state out to persistent storage so that it
+can stop and start without losing all of its data.
+
+For this tutorial, we will use [badger](https://github.com/dgraph-io/badger). 
+Badger is a fast embedded key-value store. 
+
+First, add badger as a dependency of your go module using the `go get` command:
+
+`go get github.com/dgraph-io/badger`
+
+Next, let's update the application and its constructor to receive a handle to the 
+database. 
 
-When a new transaction is added to the Tendermint Core, it will ask the
-application to check it (validate the format, signatures, etc.).
+Update the application struct as follows: 
 
 ```go
-import "bytes"
-
-func (app *KVStoreApplication) isValid(tx []byte) (code uint32) {
- // check format
- parts := bytes.Split(tx, []byte("="))
- if len(parts) != 2 {
-  return 1
- }
-
- key, value := parts[0], parts[1]
-
- // check if the same key=value already exists
- err := app.db.View(func(txn *badger.Txn) error {
-  item, err := txn.Get(key)
-  if err != nil && err != badger.ErrKeyNotFound {
-   return err
-  }
-  if err == nil {
-   return item.Value(func(val []byte) error {
-    if bytes.Equal(val, value) {
-     code = 2
-    }
-    return nil
-   })
-  }
-  return nil
- })
- if err != nil {
-  panic(err)
- }
-
- return code
+type KVStoreApplication struct {
+	db           *badger.DB
+	currentBatch *badger.Txn
 }
+```
 
-func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
- code := app.isValid(req.Tx)
- return abcitypes.ResponseCheckTx{Code: code, GasWanted: 1}
+And change the constructor to set the appropriate field when creating the application:
+
+```go
+func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
+	return &KVStoreApplication{
+		db: db,
+	}
 }
 ```
 
-Don't worry if this does not compile yet.
-
-If the transaction does not have a form of `{bytes}={bytes}`, we return `1`
-code. When the same key=value already exist (same key and value), we return `2`
-code. For others, we return a zero code indicating that they are valid.
+Don't worry about `currentBatch` for now, we'll get to that later.
```

Comment:
It's probably still worth saying what this is, at a high level, even if we don't go into detail yet. Perhaps:

> The `currentBatch` keeps track of the transactions that update the application's state when a block is completed. Don't worry about it for now, we'll go into more detail later.